jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/PerProcessInstanceRuntimeManager.java

```diff
@@ -235,34 +235,30 @@ public void disposeRuntimeEngine(RuntimeEngine runtime) {
             return;
     	}
     	try {
-        	if (canDispose(runtime)) {
-            	removeLocalRuntime(runtime);            	
-            	
-            	Long ksessionId = ((RuntimeEngineImpl)runtime).getKieSessionId();
-            	if (runtime instanceof Disposable) {
-                	// special handling for in memory to not allow to dispose if there is any context in the mapper
-                	if (mapper instanceof InMemoryMapper && ((InMemoryMapper)mapper).hasContext(ksessionId)){
-                	    logger.debug("Dispose runtime has context {} inMemoryMapper", ksessionId);
-                		return;
-                	}
+            if (runtime instanceof RuntimeEngineImpl && canDispose(runtime)) {
+                Long ksessionId = ((RuntimeEngineImpl) runtime).getKieSessionId();
+                if (runtime instanceof Disposable) {
+                    // special handling for in memory to not allow to dispose if there is any context in the mapper
+                    if (mapper instanceof InMemoryMapper && ((InMemoryMapper) mapper).hasContext(ksessionId)) {
+                        logger.debug("Dispose runtime has context {} inMemoryMapper", ksessionId);
+                        return;
+                    }
                     ((Disposable) runtime).dispose();
                 }
-                            	
-            	if (ksessionId != null) {
-                	TimerService timerService = TimerServiceRegistry.getInstance().get(getIdentifier() + TimerServiceRegistry.TIMER_SERVICE_SUFFIX);
+                if (ksessionId != null) {
+                    TimerService timerService = TimerServiceRegistry.getInstance().get(getIdentifier() + TimerServiceRegistry.TIMER_SERVICE_SUFFIX);
                     if (timerService != null) {
                         if (timerService instanceof GlobalTimerService) {
                             ((GlobalTimerService) timerService).clearTimerJobInstances(ksessionId);
                         }
                     }
-            	}
-            	logger.debug("About to release lock {}", runtime);
-            	releaseAndCleanLock(runtime);
-        	}
+                }
+            }
     	} catch (Exception e) {
-    	    releaseAndCleanLock(runtime);
-    	    removeLocalRuntime(runtime);  	    
     	    throw new RuntimeException(e);
+    	} finally {
+    	    removeLocalRuntime(runtime);
```

Comment:
Shouldn't be checked `if (canDispose(runtime))` before?

Reply:
removeLocalRuntimer should be invoked always, it is not related with dispose (you can check that in the previous version)

Reply:
@gmunozfe you were right and we need to preserve existing order for current test to work